pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -105,10 +118,88 @@ func (f *framework) RunReservePlugins(
+// running the remaining plugins and returns an error. Otherwise, if any of the
+// plugins returns "Wait", then this function will block for the timeout period
+// returned by the plugin, if the time expires, then it will return an error.
+// Note that if multiple plugins asked to wait, then we wait for the minimum
+// timeout duration.
+func (f *framework) RunPermitPlugins(
+	pc *PluginContext, pod *v1.Pod, nodeName string) *Status {
+	timeout := maxTimeout
+	statusCode := Success
+	for _, pl := range f.permitPlugins {
+		status, d := pl.Permit(pc, pod, nodeName)
+		if !status.IsSuccess() {
+			if status.Code() == Unschedulable {
+				msg := fmt.Sprintf("rejected by %v at permit: %v", pl.Name(), status.Message())
+				klog.V(4).Infof(msg)
+				return NewStatus(status.Code(), msg)
+			}
+			if status.Code() == Wait {
+				// Use the minimum timeout duration.
+				if timeout > d {
+					timeout = d
+				}
+				statusCode = Wait
+			} else {
+				msg := fmt.Sprintf("error while running %v permit plugin for pod %v: %v", pl.Name(), pod.Name, status.Message())
+				klog.Error(msg)
+				return NewStatus(Error, msg)
+			}
+		}
+	}
+
+	// We now wait for the minimum duration if at least one plugin asked to
+	// wait (and no plugin rejected the pod)
+	if statusCode == Wait {
+		w := NewWaitingPod(pod)
+		f.waitingPods.Add(pod.UID, w)
+		defer f.waitingPods.Remove(pod.UID)
+		timer := time.NewTimer(timeout)
+		klog.V(4).Infof("waiting for %v for pod %v at permit", timeout, pod.Name)
+		select {
+		case <-timer.C:
+			msg := fmt.Sprintf("pod %v rejected due to timeout after waiting %v at permit", pod.Name, timeout)
+			klog.V(4).Infof(msg)
+			return NewStatus(Unschedulable, msg)
+		case s := <-w.s:
+			if !s.IsSuccess() {
+				if s.Code() == Unschedulable {
+					msg := fmt.Sprintf("rejected while waiting at permit: %v", s.Message())
+					klog.V(4).Infof(msg)
+					return NewStatus(s.Code(), msg)
+				}
+				msg := fmt.Sprintf("error received while waiting at permit for pod %v: %v", pod.Name, s.Message())
+				klog.Error(msg)
+				return NewStatus(Error, msg)
+			}
+		}
+	}
+
+	return nil
+}
+
 // NodeInfoSnapshot returns the latest NodeInfo snapshot. The snapshot
 // is taken at the beginning of a scheduling cycle and remains unchanged until a
 // pod finishes "Reserve". There is no guarantee that the information remains
 // unchanged after "Reserve".
 func (f *framework) NodeInfoSnapshot() *cache.NodeInfoSnapshot {
 	return f.nodeInfoSnapshot
 }
+
+// IterateOverWaitingPods acquires a read lock and iterates over the WaitingPods map.
+func (f *framework) IterateOverWaitingPods(callback func(types.UID, WaitingPod)) {
+	f.waitingPods.Iterate(callback)
+}
+
+// GetWaitingPod returns a reference to a WaitingPod given its UID.
+func (f *framework) GetWaitingPod(uid types.UID) WaitingPod {
+	return f.waitingPods.Get(uid)
+}
+
+// GetWaitingPodsCount returns the number of waiting pods.
+func (f *framework) GetWaitingPodsCount() int {
```

Comment:
Let's remove this function and added when needed.

Reply:
This is actually used in the integration test.

Reply:
If it is only for testing purposes, I guess you can easily use `IterateOverWaitingPods` and pass a callback function that counts the pods.
The reason that I insist on removing this function or other functions that we don't have an obvious use for is that these functions will be a part of the framework API and will be hard to change or remove in the future. Once it is added, it will stay!